File: backend/src/crypto/providers/identity_key_provider.py
"""Grade A: ML-DSA-65 identity keys (post-quantum digital signatures, FIPS 204).

Implemented via `liboqs-python`. Primarily exercised client-side (spec
FR-051) and in crypto validation tests; the backend never generates or
holds a user's private identity key in production code paths.
"""

import oqs

from src.crypto.interfaces import IdentityKeyProvider

_ALGORITHM = "ML-DSA-65"


class MLDSA65IdentityKeyProvider(IdentityKeyProvider):
    def generate_keypair(self) -> tuple[bytes, bytes]:
        with oqs.Signature(_ALGORITHM) as signer:
            public_key = signer.generate_keypair()
            private_key = signer.export_secret_key()
            return public_key, private_key

    def sign(self, private_key: bytes, message: bytes) -> bytes:
        with oqs.Signature(_ALGORITHM, private_key) as signer:
            return bytes(signer.sign(message))

    def verify(self, public_key: bytes, message: bytes, signature: bytes) -> bool:
        with oqs.Signature(_ALGORITHM) as verifier:
            return bool(verifier.verify(message, signature, public_key))

File: frontend/src/crypto/providers/identityKeyProvider.ts
/**
 * Grade A: ML-DSA-65 identity keys (post-quantum signatures, FIPS 204),
 * via `@noble/post-quantum`. Private keys never leave this module boundary
 * unencrypted — callers persist them via the browser's local key storage
 * (IndexedDB, itself encrypted at rest by the platform), never via the
 * backend API client.
 */

import { ml_dsa65 } from "@noble/post-quantum/ml-dsa";
import type { IdentityKeyProvider } from "../interfaces";

export const mlDsa65IdentityKeyProvider: IdentityKeyProvider = {
  async generateKeyPair() {
    const seed = crypto.getRandomValues(new Uint8Array(32));
    const keys = ml_dsa65.keygen(seed);
    return { publicKey: keys.publicKey, privateKey: keys.secretKey };
  },

  async sign(privateKey, message) {
    return ml_dsa65.sign(privateKey, message);
  },

  async verify(publicKey, message, signature) {
    return ml_dsa65.verify(publicKey, message, signature);
  },
};
